File: service/python/id.py
#!/usr/bin/env python3
# -*- coding=utf-8 -*-
import os
import redis

rds = redis.Redis(
    host=str(os.environ.get('REDIS_HOST', 'localhost')),
    port=int(os.environ.get('REDIS_PORT', 6379)),
    db=0,
    password=os.environ.get('REDIS_PASSWORD', 'pernyi-awmous-leanly-crotin-shoop-cotula-embira'),
)

digit62 = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'

def int_to_str62(x):
    s = ''
    while x>0:
        s = digit62[x % 62]+s
        x = int(x/62)
    return s

def str62_to_int(s):
    x = 0
    s = str(s).strip()
    if s=="":
        return x
    for y in s:
        k = digit62.find(y)
        if k>=0:
            x=x*62+k
    return x

save_id = None

def get():
    global save_id
    if save_id is None:
        save_id = rds.incr('id')
    return int_to_str62(save_id).rjust(8, '0')

def ack():
    global save_id
    save_id = None




if __name__ == "__main__":
    print(get())
    print(get())
    ack()
    print(get())

    print(str62_to_int(int_to_str62(100000000000000).rjust(8, '0')))

File: service/node/id.js
const redis = require('redis');

const obj = {
    /**
     * @type {redis.RedisClient}
     */
    rds: null,
    id_list: [],
};
const digit62 = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

function init() {
    obj.rds = redis.createClient(
        process.env.REDIS_PORT || 6379,
        process.env.REDIS_HOST || 'localhost',
        {
            password: process.env.REDIS_PASSWORD || 'pernyi-awmous-leanly-crotin-shoop-cotula-embira'
        }
    )
}

/**
 *
 * @param {string} s
 * @param {number} width
 * @param {string} fillchar
 */
function str_rjust(s, width, fillchar) {
    const n = width - s.length;
    return fillchar.repeat(n > 0 ? n : 0) + s;
}


function int_to_str62(x) {
    let s = '';
    while(x > 0) {
        s = digit62[x%62] + s;
        x = parseInt(x/62);
    }
    return s;
}

/**
 * 获取id
 * @returns {Promise<number>}
 */
function get() {
    return new Promise((resolve)=>{
        let id = obj.id_list.shift();
        if(undefined !== id) {
            return resolve(id);
        }
        obj.rds.incr('id', (err, n)=>{
            return resolve(str_rjust(int_to_str62(n), 8, '0'));
        })
    })
}

function store(id) {
    if(obj.id_list.indexOf(id) === -1) {
        obj.id_list.push(id);
    }
}

module.exports = {
    init,
    get,
    store,
};
